statsd/aggregator.go

```diff
@@ -3,61 +3,48 @@ package statsd
 import (
 	"fmt"
 	"math"
-	"runtime"
 	"sort"
-	"strings"
-	"sync"
 	"time"
 
-	"github.com/atlassian/gostatsd/backend"
 	"github.com/atlassian/gostatsd/types"
 
 	log "github.com/Sirupsen/logrus"
 )
 
-// metricAggregatorStats is a bookkeeping structure for statistics about a MetricAggregator.
-type metricAggregatorStats struct {
-	BadLines       int64
-	LastMessage    time.Time
-	LastFlush      time.Time
-	LastFlushError time.Time
-	NumStats       int
-	ProcessingTime time.Duration
-}
+// ProcessFunc is a function that gets executed by Aggregator with its state passed into the function.
+type ProcessFunc func(*types.MetricMap)
 
-// MetricAggregator is an object that aggregates statsd metrics.
+// Aggregator is an object that aggregates statsd metrics.
 // The function NewMetricAggregator should be used to create the objects.
 //
-// Incoming metrics should be sent to the MetricQueue channel.
-type MetricAggregator struct {
-	sync.Mutex
-	ExpiryInterval    time.Duration      // How often to expire metrics
-	FlushInterval     time.Duration      // How often to flush metrics to the sender
-	LastFlush         time.Time          // Last time the metrics where aggregated
-	MaxWorkers        int                // Number of workers to metrics queue
-	MetricQueue       chan *types.Metric // Queue on which metrics are received
-	PercentThresholds []float64
-	Senders           []backend.MetricSender // The sender to which metrics are flushed
-	Stats             metricAggregatorStats
-	Tags              types.Tags // Tags to add to all metrics
+// Incoming metrics should be passed via ReceiveMetric function.
+type Aggregator interface {
+	ReceiveMetric(*types.Metric, time.Time)
+	Flush(func() time.Time) *types.MetricMap
+	Process(ProcessFunc)
+	Reset(time.Time)
+}
+
+type metricAggregator struct {
```

Comment:
And this one is a `metricAggregator` which implements `Aggregator` and has a `ReceiveMetric` function and is created via a `NewMetricAggregator` function... A bit confusing, I'd rather we are consistent.